src/wasm-binary-reader-objdump.c

```diff
@@ -0,0 +1,417 @@
+}
+
+static WasmResult on_opcode_bare(WasmBinaryReaderContext* ctx) {
+  Context* context = ctx->user_data;
+  log_opcode(context, ctx->data, 0, NULL);
+  return WASM_OK;
+}
+
+static WasmResult on_opcode_uint32(WasmBinaryReaderContext* ctx, uint32_t value) {
+  Context* context = ctx->user_data;
+  size_t immediate_len = ctx->offset - context->current_opcode_offset;
+  log_opcode(context, ctx->data, immediate_len, "%#x", value);
+  return WASM_OK;
+}
+
+static WasmResult on_opcode_uint32_uint32(WasmBinaryReaderContext* ctx, uint32_t value, uint32_t value2) {
+  Context* context = ctx->user_data;
+  size_t immediate_len = ctx->offset - context->current_opcode_offset;
+  log_opcode(context, ctx->data, immediate_len, "%lu %lu", value, value2);
+  return WASM_OK;
+}
+
+static WasmResult on_opcode_uint64(WasmBinaryReaderContext* ctx, uint64_t value) {
+  Context* context = ctx->user_data;
+  size_t immediate_len = ctx->offset - context->current_opcode_offset;
+  log_opcode(context, ctx->data, immediate_len, "%d", value);
+  return WASM_OK;
+}
+
+WasmResult on_br_table_expr(WasmBinaryReaderContext* ctx, uint32_t num_targets,
+                            uint32_t* target_depths,
+                            uint32_t default_target_depth) {
+  Context* context = ctx->user_data;
+  size_t immediate_len = ctx->offset - context->current_opcode_offset;
+  log_opcode(context, ctx->data, immediate_len, "");
+  return WASM_OK;
+}
+
+static WasmResult on_end_expr(void* user_data) {
+  Context* context = user_data;
+  context->indent_level--;
+  assert(context->indent_level >= 0);
+  log_opcode(context, NULL, 0, NULL);
+  return WASM_OK;
+}
+
+static WasmResult on_opcode_block_sig(WasmBinaryReaderContext* ctx,
+                                      uint32_t num_types,
+                                      WasmType* sig_types) {
+  Context* context = ctx->user_data;
+  log_opcode(context, ctx->data, 1, "%d", *sig_types);
+  context->indent_level++;
+  return WASM_OK;
+}
+
+static void print_wasm_type(WasmType type) {
+  switch (type) {
+    case WASM_TYPE_I32:
+      printf("i32");
+      break;
+
+    case WASM_TYPE_I64:
+      printf("i64");
+      break;
+
+    case WASM_TYPE_F32:
+      printf("f32");
+      break;
+
+    case WASM_TYPE_F64:
+      printf("f64");
+      break;
+
+    default:
+      assert(0);
+      break;
+  }
+}
+
+static WasmResult on_signature(uint32_t index,
+                               uint32_t param_count,
+                               WasmType* param_types,
+                               uint32_t result_count,
+                               WasmType* result_types,
+                               void* user_data) {
+  Context* ctx = user_data;
+
+  if (!ctx->options->verbose)
+    return WASM_OK;
+  printf(" - [%d] (", index);
+  uint32_t i;
+  for (i = 0; i < param_count; i++) {
+    if (i != 0) {
+      printf(", ");
+    }
+    print_wasm_type(param_types[i]);
+  }
+  printf(") -> ");
+  assert(result_count == 1);
```

Comment:
How does this work for result_count == 0? I'd rather just handle the multi-return signatures uniformly (since they work pretty much everywhere else).